ai21_tokenizer/jamba_instruct_tokenizer.py

```diff
@@ -44,35 +43,93 @@ def _init_tokenizer(self, model_path: PathLike, cache_dir: PathLike) -> Tokenize
 
         return tokenizer
 
-    def _is_cached(self, cache_dir: PathLike) -> bool:
-        return Path(cache_dir).exists() and _TOKENIZER_FILE in os.listdir(cache_dir)
-
     def _load_from_cache(self, cache_file: Path) -> Tokenizer:
         return cast(Tokenizer, Tokenizer.from_file(str(cache_file)))
 
-    def _cache_tokenizer(self, tokenizer: Tokenizer, cache_dir: PathLike) -> None:
-        # create cache directory for caching the tokenizer and save it
-        Path(cache_dir).mkdir(parents=True, exist_ok=True)
-        tokenizer.save(str(cache_dir / _TOKENIZER_FILE))
-
     def encode(self, text: str, **kwargs) -> List[int]:
-        return self._tokenizer.encode(text, **kwargs).ids
+        return self._encode(text, **kwargs)
 
     def decode(self, token_ids: List[int], **kwargs) -> str:
-        return self._tokenizer.decode(token_ids, **kwargs)
+        return self._decode(token_ids, **kwargs)
 
     def convert_tokens_to_ids(self, tokens: Union[str, List[str]]) -> Union[int, List[int]]:
-        if isinstance(tokens, str):
-            return self._tokenizer.token_to_id(tokens)
-
-        return [self._tokenizer.token_to_id(token) for token in tokens]
+        return self._convert_tokens_to_ids(tokens)
 
     def convert_ids_to_tokens(self, token_ids: Union[int, List[int]], **kwargs) -> Union[str, List[str]]:
-        if isinstance(token_ids, int):
-            return self._tokenizer.id_to_token(token_ids)
+        return self._convert_ids_to_tokens(token_ids)
+
+    @property
+    def vocab_size(self) -> int:
+        return self._tokenizer.get_vocab_size()
+
+
+class AsyncJambaInstructTokenizer(BaseJambaInstructTokenizer, BaseTokenizer):
+    def __init__(
+        self,
+        model_path: str,
+        cache_dir: Optional[PathLike] = None,
+    ):
+        """
+        Args:
+            model_path: str
+                The identifier of a Model on the Hugging Face Hub, that contains a tokenizer.json file
+            cache_dir: Optional[PathLike]
+                The directory to cache the tokenizer.json file.
+                 If not provided, the default cache directory will be used
+        """
+        self._model_path = model_path
+        self._cache_dir = cache_dir or _DEFAULT_MODEL_CACHE_DIR
+
+    async def __aenter__(self):
+        await self._init_tokenizer()
```

Comment:
Why are you not returning the tokenizer object from this function?

```python
self._tokenizer = await self._init_tokenizer()
```

Its not good practice to mutate a variable inside a function and depend that it had changed outside of the function

Reply:
This means you need to declare on the `_tokenizer` in the class's properties like so - 

```python
class AsyncJambaInstructTokenizer(BaseJambaInstructTokenizer, BaseTokenizer):
_tokenizer: Optional[Tokenizer] = None
```

Reply:
I declared it in BaseJambaInstructTokenizer, because it's a variable that both sync and async classes have